policies/codingstyle.txt

```diff
@@ -289,12 +289,81 @@ options.
 
 In source files, separate functions with one blank line.
 
+                Chapter 6.1: Function Arguments
+
+Most API functions in OpenSSL revolve around a specific type (e.g. SSL_CTX,
+EVP_PKEY, X509 etc). When adding new API functions relevant to a given type, the
+following argument ordering should be used:
+
+    rettype TYPE_get_something(TYPE *,[args])
+    rettype TYPE_set_something(TYPE *,[args])
+    rettype TYPE_do_something(TYPE *,[args])
+    TYPE *TYPE_new([args])
+
+Some more complex complex APIs have a TYPE and TYPE_CTX pair, where TYPE usually
+represents the set of function pointers to the implementation, while TYPE_CTX is
+the collection of attributes to perform the intended operation.
+For that case, the pattern should also include:
+
+    TYPE_CTX *TYPE_CTX_new([args])
+    rettype TYPE_CTX_get_something(TYPE_CTX *,[args])
+    rettype TYPE_CTX_set_something(TYPE_CTX *,[args])
+    rettype TYPE_do_something(TYPE_CTX *,[args])
+
```

Comment:
If you are adding in TYPE_CTX here then you really do need to add in the TYPE_METHOD pattern which is actually explicitly this concept and much more common in the library.

Reply:
Is that really necessary? METHOD seems to be something we're moving away from - at least in 3.0 we've deprecated several and not added any new ones that I recall. In terms of public APIs for creating METHODs that aren't already deprecated I only see `BIO_meth_new`, `X509_LOOKUP_meth_new` and `X509_CRL_METHOD_new`. It doesn't seem as prevalent a concept as TYPE_CTX (in the public API)

Reply:
> > So, as factories (pools) go, it seems clear that we have two, SSL_CTX and OPENSSL_CTX
> I don't view OPENSSL_CTX as a factory. The only ctx as a factory I can think of is SSL_CTX.

The fun of not having a common vocabulary. 

There are multiple types of factories in OO programming and OPENSSL_CTX holds what is necessary for creating new instances of a variety of different types (of object). It is a factory. That is what factories do. You use factories for a variety of different purposes.

Reply:
> METHOD seems to be something we're moving away from

I believe that's accidental rather than intentional.  Most of our focus has been on EVP, where the TYPE_CTX / TYPE naming is more prevalent...  the question, however, is if it's *right*, and that can be measure on the "what's confusing even to us" scale.  In that regard, I find EVP *quite* confusing, especially, since it's a mix of both patterns (see EVP_PKEY / EVP_PKEY_ASN1_METHOD and EVP_PKEY_CTX / EVP_PKEY_METHOD).

Reply:
(and before you say it, I fell into the TYPE_CTX / TYPE pattern as well...  I can clearly see how EVP affected me :confused:)

Reply:
> I believe that's accidental rather than intentional.

Perhaps, but it seems to me that the new provider way of doing things largely replaces a lot of the places where we might have used METHODs before. I don't feel a strong need to incorporate that pattern in the coding style.

Reply:
Point is, what patterns do we want to see going forward, and I at least find the TYPE / TYPE_METHOD much clearer than the style of TYPE_CTX / TYPE that we have in EVP (and that very unfortunately bled a bit elsewhere, something I'll get to regret quite a bit)

Reply:
@mattcaswell asked me to try and describe the patterns that have been identified, I do it with a diff.  I purposefully included a constructor that uses `OSSL_LIB_CTX` as a context that refers to a factory pool (to not have that dispute again, *indirectly*...  it doesn't really matter to the user, as @t-j-h has pointed out):

``` diff
diff --git a/policies/codingstyle.txt b/policies/codingstyle.txt
index 6a1e5aa..9ffaab3 100644
--- a/policies/codingstyle.txt
+++ b/policies/codingstyle.txt
@@ -300,9 +300,20 @@ following argument ordering should be used:
     rettype TYPE_do_something(TYPE *,[args])
     TYPE *TYPE_new([args])
 
-Some more complex complex APIs have a TYPE and TYPE_CTX pair, where TYPE usually
-represents the set of function pointers to the implementation, while TYPE_CTX is
-the collection of attributes to perform the intended operation.
+Some more complex APIs with diverse underlying implementations have TYPE
+paired with a TYPE_METHOD, where TYPE usually represents a collection of
+attributes while TYPE_METHOD represents the set of function pointers to the
+implementation. The public API remains as above, except for the constructor,
+which should look like one of these (the latter being a constructor that
+uses a factory pool referred to by some context):
+
+    TYPE *TYPE_new(TYPE_METHOD *, [args])
+    TYPE *TYPE_new(OSSL_LIB_CTX *, TYPE_METHOD *, [args])
+
+A variant for the same sort of complex APIs is also seen, with the
+corresponding type pair is TYPE_CTX and TYPE, where TYPE usually
+represents the set of function pointers to the implementation, while
+TYPE_CTX is the collection of attributes to perform the intended operation.
 For that case, the pattern should also include:
 
     TYPE_CTX *TYPE_CTX_new([args])
```

Reply:
I can live with that.